### DISCUSSION TITLE
How to make VSC Solidity lint recognize your @chainlink and @openzeppelin libraries

### DISCUSSION THREAD

**Discussion by n4n0b1t3** [smartcontractkit/full-blockchain-solidity-course-py#1069] [General]
The following solution is written for Windows, but Mac should be very similar with minor changes in the path.
Please keep in mind that lint it is not necessary for the code to work, this is just for lint to recognize the libraries.

1. Please check if you have the desired packages already installed with `brownie pm list`
2. If the package is not listed you will need to install it first. Here an example for OpenZeppelin 
`brownie pm install OpenZeppelin/openzeppelin-contracts@3.0.0`
3. Regardless of whether you installed the package under a virtual environment, Brownie will store the package as follows.
`C:\Users\Username\.brownie\packages ...`
or more specific in 
`C:\Users\Zarathustra\.brownie\packages\OpenZeppelin\openzeppelin-contracts@3.0.0`
or here is the path for chainlink
`C:\Users\Zarathustra\.brownie\packages\smartcontractkit\chainlink-brownie-contracts@0.3.0`

4. Copy the path, for Windows remove the drive letter and revert the backslashes into (forwad) slashes, the Chainlink path should now look like this:
`/Users/Zarathustra/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.3.0`

5. In Visual Studio Code press `Ctrl + ,` and type into the search box `solidity.remap`
![image](https://user-images.githubusercontent.com/32838961/154129048-9ddffcb2-5309-4875-98b0-a3964d20d01f.png)

6. Click the small link (see screenshot) and open the settings file

Adapt the following lines to your needs and add them to the settings file.
```
	,
    "solidity.remappings": [
        "@chainlink/=/Users/UserName/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.3.0",
    ],
```

The logic is as follows: Whenever VSC lint encounters "@chainlink/" it will seach for the library in the path which is written after the equal sign. So @chainlink/contracts will look at the path and add /contracts etc.

You can add more solidity.remappings into the data structure separated by commas and enclosed in double quotation marks.

Here just for comparison my settings.json file (do not blindly copy it)
```
{
    "editor.suggestSelection": "first",
    "vsintellicode.modify.editor.suggestSelection": "automaticallyOverrodeDefaultValue",
    "python.pythonPath": "C:\\Users\\UserName\\pythonversions\\python3.9.1\\python.exe",
    "python.languageServer": "Pylance",
    "python.venvFolders": [
        "C:\\Users\\UserName\\pythonprojects",
        "C:\\Users\\UserName\\blockchainProjects"
    ],
    "workbench.editorAssociations": {
        "*.ipynb": "jupyter.notebook.ipynb"
    },
    "python.defaultInterpreterPath": "C:\\Users\\UserName\\pythonversions\\python3.9.1\\python.exe",
    "solidity.packageDefaultDependenciesDirectory": "node_modules",
    "solidity.compileUsingRemoteVersion": "latest",
    "solidity.remappings": [
        "@chainlink/=/Users/UserName/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.3.0",
    ],
    "editor.formatOnSave": true,
    "python.formatting.provider": "black",
    "editor.bracketPairColorization.enabled": true,
    "python.venvPath": "~/.venv"
}
```

The changes should take effect immediately.

**Comment by cromewar** [comment]
Please everyone leave a clap on [Medium](https://n4n0b1t3.medium.com/how-to-make-vsc-solidity-lint-recognize-your-chainlink-and-openzeppelin-libraries-73775129261c) for @n4n0b1t3 and this wonderful article!

**Reply by ofirbenatar** [reply]
gave him 5 clap, he just saved me a lot of time + taught me how to manually add paths so vscode will recognize packages.  @PatrickAlphaC, you, @n4n0b1t3, and all the other contributors are doing a god work for a young guy and his family living in the middle east who is just trying to move himself and his loves one from zero to one in life thank you
